### PR TITLE
[Refactoring] Rely on Streams in Mailbox APIs

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by Arsnael** [comment]
CI restart...

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[8f726a48261083e9511cb8672507e638695ab633] [ERROR] Errors: 
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.deleteMessageShouldNotRemoveMessageOnAnotherMailbox:320 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.deleteMessageShouldRemoveMessageOnlyFromMailbox:299 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxShouldAllowDistinctMailboxSetForShareeAndOwner:848 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldAllowCopyingMessageFromReadOnlySharedMailbox:669 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldAllowCopyingToAInsertSharedMailbox:725 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldAllowMovingMessagesFromASharedMailboxWhenDeleteRight:785 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldDenyCopyingToANonInsertSharedMailbox:757 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldDenyMovingMessagesFromASharedMailboxWhenNoDeleteRight:817 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerStorageTest>AbstractMessageIdManagerStorageTest.setInMailboxesShouldPreserveMessageFromOneMailboxDeletion:534 » IllegalState
[8f726a48261083e9511cb8672507e638695ab633] [ERROR]   InMemoryMessageIdManagerSideEffectTest>AbstractMessageIdManagerSideEffectTest.setFlagsShouldDispatchWhenMessageBelongsToTheMailboxes:463 » IllegalState
```

**Comment by Arsnael** [comment]
```
[bc0d66e1b60b266ecb4873e4c1ee548fd337f53c] [ERROR]   InMemoryMessageIdManagerSideEffectTest>AbstractMessageIdManagerSideEffectTest.setFlagsShouldDispatchWhenMessageBelongsToTheMailboxes:463 » IllegalState
```

**Comment by Arsnael** [comment]
Need rebase @chibenwa

**Comment by chibenwa** [comment]
```
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.multimailboxSearchShouldReturnUidOfMessageMarkedAsSeenInAllMailboxes:551 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.multimailboxSearchShouldReturnUidOfMessageMarkedAsSeenInOneMailbox:560 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.multimailboxSearchShouldReturnUidOfMessageMarkedAsSeenInTwoMailboxes:586 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.multimailboxSearchShouldReturnUidOfMessageWithExpectedFromInTwoMailboxes:573 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.multimailboxSearchShouldWorkWithOtherUserMailbox:616 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.searchingMessageInMultipleMailboxShouldNotReturnLessMessageThanLimitArgument:321 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.searchingMessageInMultipleMailboxShouldNotReturnLessMessageThanLimitArgumentEvenIfDuplicatedMessageAreBeforeLegitimeMessage:358 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.searchingMessageInMultipleMailboxShouldNotReturnTwiceTheSameMessage:253 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.searchingMessageInMultipleMailboxShouldUnionOfTheTwoMailbox:285 » Completion
[cb0bfe8746428da2b226785af2cd7968fe15b235] [ERROR]   ElasticSearchIntegrationTest>AbstractMessageSearchIndexTest.sortShouldNotDiscardResultWhenSearchingFieldIsIdentical:1410 » Completion
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java [hunk 1]
```diff
@@ -219,6 +219,8 @@
 
     /**
      * Searches for mailboxes matching the given query.
+     *
+     * Caller is responsible to close this stream to ensure resources and search context get released.
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: Why this warning here and not on every other methods?

**Comment by Arsnael** [comment]
```
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java:[88,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java:[116,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java:[115,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java:[149,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageMetadataViewFactoryTest.java:[81,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
[13cda15a71d22dd85a1d4f1a9a2f5df5df265113] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageMetadataViewFactoryTest.java:[100,25] incompatible types: java.util.stream.Stream<org.apache.james.mailbox.model.MessageResult> cannot be converted to java.util.List<org.apache.james.mailbox.model.MessageResult>
```

**Comment by chibenwa** [comment]
Well, I must admit I concidered the streams as an intermediate for a reactive migration to be honest.

Would it makes sense to you @mbaechler ?

**Comment by Arsnael** [comment]
```
[2df0fdcf66529a151018c2077c8e591f0f728246] [ERROR] Errors: 
[2df0fdcf66529a151018c2077c8e591f0f728246] [ERROR]   FileMailQueueTest.dequeueShouldBeConcurrent » ConditionTimeout Condition with ...
```
test this please

**Comment by mbaechler** [comment]
> Well, I must admit I concidered the streams as an intermediate for a reactive migration to be honest.
> 
> Would it makes sense to you @mbaechler ?

I fear we'll have to spend time chasing resource leaks with this solution and given that Reactor is not much harder to implement, I would really opt for Publisher.

However, if we are confident about resource leaks, we can adopt that PR

**Comment by chibenwa** [comment]
Let's get it green first...

**Comment by chibenwa** [comment]
> However, if we are confident about resource leaks, we can adopt that PR

From my knowledge ES is the only place we rely on that stream feature.... Other places are streams from in memory data sources.

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.imapMovesToSpamMailboxShouldBeConsideredAsSpam » ConditionTimeout`

test this please

**Comment by Arsnael** [comment]
```
[2df0fdcf66529a151018c2077c8e591f0f728246] [ERROR]   FileMailQueueTest.dequeueShouldBeConcurrent » ConditionTimeout Condition with ...
```
test this please

**Comment by chibenwa** [comment]
`FileMailQueueTest.dequeueShouldBeConcurrent » ConditionTimeout Condition with ...`

test this please

**Comment by Arsnael** [comment]
Need rebase

**Comment by chibenwa** [comment]
Remove milestone as I will not handle that in the coming week.

**Comment by chibenwa** [comment]
Was :green_apple:

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java (hunk 1 above)
Comment (rouazana) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: not fixed?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java [hunk 2]
```diff
@@ -118,7 +117,7 @@
      * @throws MailboxException
      *             if anything went wrong
      */
-    Iterator<MessageUid> expunge(MessageRange set, MailboxSession mailboxSession) throws MailboxException;
+    Stream<MessageUid> expunge(MessageRange set, MailboxSession mailboxSession) throws MailboxException;
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: also here?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 3]
```diff
@@ -86,7 +87,7 @@ public CassandraMessageIdMapper(MailboxMapper mailboxMapper, CassandraMailboxDAO
     }
 
     @Override
-    public List<MailboxMessage> find(Collection<MessageId> messageIds, FetchType fetchType) {
+    public Stream<MailboxMessage> find(Collection<MessageId> messageIds, FetchType fetchType) {
```
[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: not answered?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: See https://github.com/linagora/james-project/pull/2919#issuecomment-558482715

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: See https://github.com/linagora/james-project/pull/2919#issuecomment-558482715

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: https://github.com/linagora/james-project/pull/2919#issuecomment-558549766

> > However, if we are confident about resource leaks, we can adopt that PR

https://github.com/linagora/james-project/pull/2919#issuecomment-558552465

> From my knowledge ES is the only place we rely on that stream feature.... Other places are streams from in memory data sources.

And the minimum we can do is document this specific behaviour ;-)

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: What do you mean?

Comment (rouazana) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: it was about adding the comment on stream.close

**Comment by Arsnael** [comment]
Rebased for solving conflicts with master branch

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**